[JsonPath] Add the component

alexandre-daubois opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #57280
| License       | MIT

## JsonPath component

Today I'm presenting the JsonPath component. Thanks to this component, it will be possible to query JSON strings using the JSON Path syntax, as described in the recent RFC 9535. This RFC was released in February 2024 and is published here: https://datatracker.ietf.org/doc/html/rfc9535.

Here's a preview of what's possible:

```php
<?php

require 'vendor/autoload.php';

use Symfony\Component\JsonPath\JsonPath;
use Symfony\Component\JsonPath\JsonCrawler;

$json = <<<'JSON'
{"store": {"book": [
    {"category": "reference", "author": "Nigel Rees", "title": "Sayings", "price": 8.95},
    {"category": "fiction", "author": "Evelyn Waugh", "title": "Sword", "price": 12.99}
]}}
JSON;

$crawler = new JsonCrawler($json);

// Basic property access
$result = $crawler->find('$.store.book[0].title');
// Array slice
$result = $crawler->find('$.store.book[0:2]');
// Reverse array slice
$result = $crawler->find('$.store.book[::-1]');
// Filter expression
$result = $crawler->find('$.store.book[?(@.price < 10)]');
// Recursive descent
$result = $crawler->find('$..author');

// Call to "length()"
$result = $crawler->find('$.store.book[?length(@.author) > 11]');

// Call to "match()"
$result = $crawler->find('$.store.book[?match(@.author, "[A-Z].*el.+")]');

// use a builder to create your path
$path = new JsonPath();
$path = $path->key('book')
    ->index(0)
    ->key('author');

$result = $crawler->find($path);
```

As stated in RFC 9535, this component embeds a few read-to-use functions:

- `length`
- `count`
- `value`
- `search`
- `match`

### Integration of JsonStreamer for performance

Thanks to the powerfulness of JsonStreamer's Splitter, we're able to guess which part of the JSON needs and doesn't need to be decoded. Querying a whole node means we can only `json_decode()` the substring containing the node. The same goes when encountering array keys. We extract the relevant substring thanks to Splitter that provides an API for that. It brings support for JSON-as-resource evaluation.

We stop when we encounter an operation that implies filtering as we need to decode the whole node to filter children.

### What about other packages?

A few alternatives exist, however they were not updated in the last few years. If they are up-to-date, they do not seem to be following RFC 9535, but rather a partial implementation of https://goessner.net/articles/JsonPath/.

### Why not include it in DomCrawler or PropertyAccess?

That was my first thought, however DomCrawler and JsonPath actually share absolutely no logic and their purpose is really different. As they have not much in common, that would be no ideal to tie them up.

PropertyAccess could be another possibility, here's where I explained why I think this would not be a correct fit: https://github.com/symfony/symfony/pull/59655#issuecomment-2628923623.

### Does it need external dependencies?

No! This component is written in vanilla PHP and doesn't require any third-party package to work.

### How will it be leveraged in Symfony/PHP?

So many possibilities :wink: The first I can think of is leveraging this component in integration tests.

Indeed, this would allow to easily validate and write new assert methods when writing integration tests with Symfony. Validating JSON returned by an API thanks to this notation would be way easier (and more readable). In short, HttpClient, BrowserKit and `WebTestCase` could beneficiate from this (especially `BrowserKitAssertionsTrait` where asserts on Json could be added!).

Symfony would not be the only beneficiary: we can easily imagine that libraries like Behat (or PHPUnit, why not) could use this package to implement asserts on JSON.

Apart from testing frameworks, this package can also be used to quickly extract precise data from a JSON. So, possibilities here are also endless.

### Possible evolution

- Leverage ExpressionLanguage to evaluate filters and allow user to inject its own filter functions (custom functions are allowed per the RFC)
- Improve the JsonPath builder
- XPath may be converted to JSON Path fairly easily, this could be something to investigate on

alexandre-daubois committed:
[JsonPath] Add the component as experimental

OskarStark commented on src/Symfony/Component/JsonCrawler/Exception/InvalidJsonException.php:
```diff
@@ -0,0 +1,16 @@
+
+namespace Symfony\Component\JsonCrawler\Exception;
+
+class InvalidJsonException extends \RuntimeException
```
Can you describe this exception, because if it is really about invalid json, it should extend `\InvalidArgumentException`, no?

alexandre-daubois replied:
Indeed, fixed this part. Thanks 👍

mvhirsch commented:
> Symfony would not be the only beneficiary: we can easily imagine that libraries like Behat (or PHPUnit, why not) could use this package to implement asserts on JSON.

Not to forget `zenstruck/browser` which has [opt-in support for jmespath](https://github.com/zenstruck/browser?tab=readme-ov-file#json-assertions). Having native support for JSONPath might be awesome. WDYT @kbond?

hantsy commented:
> As stated in RFC 9535, this component embeds a few read-to-use functions:

Add the raw `jsonPath` function to evaluate the JsonPath expression directly.

Kocal commented on src/Symfony/Component/JsonCrawler/JsonCrawler.php:
```diff
@@ -0,0 +1,398 @@
+        }
+    }
+
+    public function find(string|JsonPath $query): array
```
It can be interesting to enforce return type to `array<array|string|float|int|bool|null>`

alexandre-daubois replied:
Definitely! I added the relevant `@return`, thanks 👍

Kocal reviewed (commented):
Cool!

kbond commented:
> Not to forget zenstruck/browser which has opt-in support for jmespath. Having native support for JSONPath might be awesome. WDYT @kbond?

Absolutely! That's the first thing I thought of.

stof commented on src/Symfony/Component/JsonCrawler/JsonCrawler.php:
```diff
@@ -0,0 +1,401 @@
+        try {
+            $this->data = json_decode($json, true, 512, \JSON_THROW_ON_ERROR);
+        } catch (\JsonException $e) {
+            throw new InvalidInputJsonException($e->getMessage());
```
I suggest attaching `$e` as previous exception as well.

stof commented on src/Symfony/Component/JsonCrawler/JsonCrawler.php:
```diff
@@ -0,0 +1,401 @@
+        } catch (InvalidJsonPathException $e) {
+            throw $e;
+        } catch (\Throwable $e) {
+            throw new JsonCrawlerException($query, $e->getMessage());
```
`$e` should be attached as previous exception.

stof commented on src/Symfony/Component/JsonCrawler/Tokenizer/JsonPathTokenizer.php:
```diff
@@ -0,0 +1,171 @@
+final class JsonPathTokenizer
+{
+    /**
+     * @return array<JsonPathToken>
```
We generally use `JsonPathToken[]` to write a type for an array with any kind of array key (both notations are strictly equivalent in term of what type they represent)

stof commented on src/Symfony/Component/JsonCrawler/Tokenizer/JsonPathTokenizer.php:
```diff
@@ -0,0 +1,171 @@
+        $quoteChar = '';
+        $filterParenthesisDepth = 0;
+
+        $chars = str_split((string) $query);
```
is a JSON path required to be ASCII only ? Or could identifiers be using multibyte Unicode characters ?

alexandre-daubois replied:
You're right. It now uses `mb_str_split` and the Tokenizer now better supports and validate the right set of character.

stof commented on src/Symfony/Component/JsonCrawler/JsonCrawler.php:
```diff
@@ -0,0 +1,401 @@
+    }
+
+    /**
+     * @return array<array|string|float|int|bool|null>
```
This phpdoc should actually be put in the interface rather than the implementation (we don't need to duplicate it for the implementation as it will inherit it)

alexandre-daubois replied:
I did so by thinking that developers could implement the interface and may want to return an array of something else, like objects with deserialization happening in their implementation or something like that. I don't have a strong opinion here, I'm fine putting it in the interface if you're still convinced that it belongs to the interface 👍

stof replied:
Well, an interface should be describing its contract precisely, otherwise caller of the interface won't know what they should handle as value being returned.

dunglas commented:
Why "JsonCrawler" and not "JsonPath"?

stof commented on src/Symfony/Component/JsonCrawler/JsonCrawler.php:
```diff
@@ -0,0 +1,401 @@
+     */
+    public function find(string|JsonPath $query): array
+    {
+        $this->evaluatedPath = \is_string($query) ? new JsonPath($query) : $query;
```
why storing such a state ?

alexandre-daubois replied:
Removed, not needed anymore actually

stof commented on src/Symfony/Component/JsonCrawler/JsonCrawler.php:
```diff
@@ -0,0 +1,401 @@
+            },
+            'count' => \is_array($value) ? \count($value) : 0,
+            'match' => match (true) {
+                \is_string($value) && \is_string($args[1] ?? null) => (bool) @preg_match("/{$args[1]}/", $value),
```
should we really silence errors for invalid patterns ?

alexandre-daubois replied:
I would say yes, per [this section of the RFC](https://www.rfc-editor.org/rfc/rfc9535.html#section-2.4.6) about the `match` function:

> If the first argument is not a string or the second argument is not a string conforming to [[RFC9485](https://www.rfc-editor.org/rfc/rfc9535.html#RFC9485)], the result is LogicalFalse

We could throw something to give a better DX, but it would not follow the specification, if I understood it correctly.

alexandre-daubois commented:
> Why "JsonCrawler" and not "JsonPath"?

Had the same thought earlier today, happy to see I'm not the only one! Let's see if anybody's against this naming. Otherwise, let's call it JsonPath, makes more sense to me as well.

GromNaN commented on src/Symfony/Component/JsonCrawler/JsonCrawler.php:
```diff
@@ -0,0 +1,401 @@
+
+        // string literals
+        if (preg_match('/^([\'"])(.*)\1$/', $expr, $matches)) {
+            return $matches[2];
```
I think it's necessary to unescape the literal string.
> Double-quoted strings follow the JSON string syntax ([Section 7](https://rfc-editor.org/rfc/rfc8259#section-7) of [[RFC8259](https://www.rfc-editor.org/rfc/rfc9535.html#RFC8259)]);
https://www.rfc-editor.org/rfc/rfc9535.html#name-syntax-3

alexandre-daubois replied:
Did so and added a test about it. Thanks!

GromNaN commented on src/Symfony/Component/JsonCrawler/JsonCrawler.php:
```diff
@@ -0,0 +1,401 @@
+            }
+            $result = [];
+            foreach ($indices as $index) {
+                $index = (int) $index;
```
The code can be simplified by removing `$indices`.
```suggestion
            foreach (explode(',', $expr) as $index) {
                $index = (int) trim($index);
```

alexandre-daubois replied:
Indeed, thanks!

GromNaN reviewed (commented):
I also think that JsonPath would be more appropriate as this is a well-known name of the spec that is implemented.

But maybe it should be part of the **PropertyAccess**? component. Why restrict this rich syntax to JSON documents when you can browse a tree of PHP objects in the same way. It would give more power to the PropertyAccess component.

alexandre-daubois commented:
@GromNaN I think that's a good question. Here's why I'm not 100% sure that would be a fit.

> Why restrict this rich syntax to JSON documents when you can browse a tree of PHP objects in the same way

Given the complexity of PropertyAccess (the logic to know if a property is writable/readable, get the value, read it, etc. is far from being trivial), I'm afraid it would bring a lot more complexity to JsonPath. The few times I fixed something in PropertyAccess was a whole challenge by itself just to understand all supported cases. **However**, what would be possible is to create something like a converter in a second time, to convert a JsonPath to a PropertyPath. PropertyAccess could then have an opt-in possibility to be used together with JsonPath.

> It would give more power to the PropertyAccess component

Definitely, but wouldn't it be too much power actually? PropertyAccess is, again, very complex and supports _a lot_ of cases. But let's put implementation complexity aside. In my opinion, it would be very confusing for developers to have, in the same component, two different syntaxes that look alike, but doesn't have the same possibilities at all. For newcomers, it could be really puzzling.

Finally, if external packages (like testing frameworks and the ones I mentioned in the description) would like to support JsonPath, I'm not sure we shouldn't force to install the whole PropertyAccess component with it.

But I still think that having an opt-in dependency to JsonPath in PropertyAccess, to convert from one syntax to another, is something that may be investigated!

xabbuh commented on src/Symfony/Component/JsonCrawler/README.md:
```diff
@@ -0,0 +1,13 @@
+JsonCrawler Component
+=====================
+
+The JsonCrawler component eases JSON navigation by manipulating JSON path as described in RFC 9535.
```
```suggestion
The JsonCrawler component eases JSON navigation by manipulating JSON paths as described in RFC 9535.
```

Like this? And maybe also link to the RFC?

xabbuh commented on src/Symfony/Component/JsonCrawler/Exception/JsonCrawlerException.php:
```diff
@@ -0,0 +1,20 @@
+
+namespace Symfony\Component\JsonCrawler\Exception;
+
+class JsonCrawlerException extends \RuntimeException
```
missing `@experimental` annotation

xabbuh commented on src/Symfony/Component/JsonCrawler/Exception/InvalidJsonPathException.php:
```diff
@@ -0,0 +1,20 @@
+
+namespace Symfony\Component\JsonCrawler\Exception;
+
+class InvalidJsonPathException extends \RuntimeException
```
missing `@experimental` annotation

xabbuh commented on src/Symfony/Component/JsonCrawler/Exception/InvalidInputJsonException.php:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Thrown when a JSON passed as an input is invalid, e.g. in {@see JsonCrawler}.
+ */
+class InvalidInputJsonException extends \InvalidArgumentException
```
missing `@experimental` annotation

alexandre-daubois commented:
Thank you @GromNaN and @stof for your review, I addressed all comments and renamed the component JsonPath.

xabbuh commented on src/Symfony/Component/JsonPath/Exception/InvalidInputJsonException.php:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Thrown when a JSON passed as an input is invalid, e.g. in {@see JsonCrawler}.
+ */
+class InvalidInputJsonException extends \InvalidArgumentException
```
missing `@experimental` annotation

xabbuh commented on src/Symfony/Component/JsonPath/Exception/InvalidJsonPathException.php:
```diff
@@ -0,0 +1,20 @@
+
+namespace Symfony\Component\JsonPath\Exception;
+
+class InvalidJsonPathException extends \RuntimeException
```
missing `@experimental` annotation

xabbuh commented on src/Symfony/Component/JsonPath/Exception/JsonCrawlerException.php:
```diff
@@ -0,0 +1,20 @@
+
+namespace Symfony\Component\JsonPath\Exception;
+
+class JsonCrawlerException extends \RuntimeException
```
missing `@experimental` annotation

xabbuh commented on src/Symfony/Component/JsonPath/README.md:
```diff
@@ -0,0 +1,13 @@
+JsonPath Component
+==================
+
+The JsonPath component eases JSON navigation by manipulating JSONPath as described in RFC 9535.
```
```suggestion
The JsonPath component eases JSON navigation by manipulating JSON paths as described in RFC 9535.
```

Like this? And maybe also link to the RFC?

alexandre-daubois replied:
I added the link to the RFC, good idea.

The RFC describes the "JSONPath syntax", so I tried an alternate version: `The JsonPath component eases JSON navigation using the JSONPath syntax`. What do you think?

xabbuh replied:
👍 this reads better

xabbuh commented on src/Symfony/Component/JsonPath/composer.json:
```diff
@@ -0,0 +1,29 @@
+{
+    "name": "symfony/json-path",
+    "type": "library",
+    "description": "Eases JSON navigation by manipulating JSONPath as described in RFC 9535.",
```
```suggestion
    "description": "Eases JSON navigation by manipulating JSON paths as described in RFC 9535.",
```

Like this?

GromNaN commented:
Thanks for exploring the proposal for `PropertyAccess`. I agree that the usage case is very different, and that it would add a lot of complexity.

dunglas commented:
Open question: would it be possible in some way to ruse the parsing infrastructure provider by JsonEncoder, so we wouldn't have to decode the whole JSON document and store it in memory to access a path. This would provide a huge performance improvement.

mtarld commented on src/Symfony/Component/JsonPath/Exception/InvalidInputJsonException.php:
```diff
@@ -0,0 +1,25 @@
+{
+    public function __construct(string $message, ?\Throwable $previous = null)
+    {
+        parent::__construct(\sprintf('Invalid input JSON: %s.', $message), previous: $previous);
```
```suggestion
        parent::__construct(\sprintf('Invalid input JSON: "%s".', $message), previous: $previous);
```

alexandre-daubois replied:
I'm not sure. This would result in: `Invalid input JSON: "unexpected characters".` for example. We should rather keep the quotes when having the path in the exception, what do you think?

(Same for the second comment)

mtarld commented on src/Symfony/Component/JsonPath/Exception/InvalidJsonPathException.php:
```diff
@@ -0,0 +1,23 @@
+{
+    public function __construct(string $message, ?int $position = null)
+    {
+        parent::__construct(\sprintf('JSONPath syntax error%s: %s', $position ? ' at position '.$position : '', $message));
```
```suggestion
        parent::__construct(\sprintf('JSONPath syntax error%s: "%s"', $position ? ' at position '.$position : '', $message));
```

mtarld commented on src/Symfony/Component/JsonPath/Exception/InvalidInputJsonException.php:
```diff
@@ -0,0 +1,25 @@
+ *
+ * @experimental
+ */
+class InvalidInputJsonException extends \InvalidArgumentException
```
I think we can create a `Symfony\Component\JsonPath\Exception\InvalidArgumentException`, implementing a  `Symfony\Component\JsonPath\Exception\ExceptionInterface`, and make the `InvalidInputJsonException` extend it. In that way, it'll be possible to easily catch JsonPath related exceptions.

alexandre-daubois replied:
Here you go 👍

mtarld commented on src/Symfony/Component/JsonPath/Exception/InvalidJsonPathException.php:
```diff
@@ -0,0 +1,23 @@
+/**
+ * @experimental
+ */
+class InvalidJsonPathException extends \RuntimeException
```
Shouldn't it be a logic exception instead?

mtarld commented:
I totally agree with @dunglas, and I even wonder if it should be a subset of the JsonEncoder component, as it will perfectly fit in that component scope and can leverage easily its lexer/splitter (which are internal at the moment).

yceruto commented on src/Symfony/Component/JsonPath/README.md:
```diff
@@ -0,0 +1,13 @@
+==================
+
+The JsonPath component eases JSON navigation using the JSONPath syntax as described in [RFC 9535](https://www.rfc-editor.org/rfc/rfc9535.html).
+
```
could we also add a simple example of how to use the component?

yceruto replied:
I see other components doing it in the README.md file directly

alexandre-daubois replied:
Good idea, added a few lines and did the same as e.g. the RateLimiter README

alexandre-daubois commented:
> But maybe this implies a too big scope for the ones who just want to use JSON pointers

Indeed, JsonEncoder is already complex and I fear that would be the first step to get back to the Serializer component capable of everything

---

As discussed with @mtarld, the Splitter could be great help. For the PR purpose, I removed the `@internal` annotation from it, maybe this should be done is a separate PR.

Putting this aside, thank you for the idea @dunglas! Happy to say that I achieved the goal. I just introduced `JsonPathTokenReducer` (name to be discussed, not 100% happy of it) which takes advantage of JsonEncoder's `Splitter`. It finds the smallest deserializable part of the JSON string given the JSONPath. It then substrings the JSON and adapt the JSONPath accordingly. I think it works great, I wrote a bunch of tests for it. The rest of the test suite is untouched after the introduction of this class. Tell me what you think! 🙂

mtarld commented on src/Symfony/Component/JsonPath/JsonCrawler.php:
```diff
@@ -0,0 +1,402 @@
+    ];
+
+    public function __construct(
+        private readonly string $raw,
```
This must handle resources as well to be compatible with streaming I guess.

mtarld commented on src/Symfony/Component/JsonPath/Tokenizer/JsonPathTokenReducer.php:
```diff
@@ -0,0 +1,75 @@
+     *
+     * @return array{json: string, tokens: list<JsonPathToken>}
+     */
+    public static function findSmallestDeserializableStringAndPath(array $tokens, string $json): array
```
The JSON must be a resource in the first place, otherwise we won't have any gain in terms of memory

mtarld commented on src/Symfony/Component/JsonPath/Tokenizer/JsonPathTokenReducer.php:
```diff
@@ -0,0 +1,75 @@
+                break;
+            }
+
+            if (!\array_key_exists($token->value, $boundaries) || \count($remainingTokens) <= 1) {
```
I'd rather avoid the `iterator_to_array` and use a `foreach`/`break`/`return` to improve the memory usage.

mtarld commented on src/Symfony/Component/JsonEncoder/Decode/Splitter.php:
```diff
@@ -17,8 +17,6 @@
  *
  * @author Mathias Arlaud <mathias.arlaud@gmail.com>
- *
- * @internal
```
`@experimental`

mtarld commented on src/Symfony/Component/JsonPath/JsonCrawler.php:
```diff
@@ -0,0 +1,402 @@
+    private function evaluate(JsonPath $query): array
+    {
+        try {
+            $simplified = JsonPathTokenReducer::findSmallestDeserializableStringAndPath(
+                JsonPathTokenizer::tokenize($query),
+                $this->raw,
+            );
```
This will require `symfony/json-encoder`. I'd make this feature opt-in if the package is installed.
And if the `symfony/json-encoder` is not installed and `$this->raw` is a resource, I'll have thrown a meaningful exception telling to install the package.

alexandre-daubois replied:
That was indeed missing, thanks!

mtarld commented on src/Symfony/Component/JsonPath/JsonCrawlerInterface.php:
```diff
@@ -0,0 +1,25 @@
+interface JsonCrawlerInterface
+{
+    /**
+     * @return array<array|string|float|int|bool|null>
```
```suggestion
     * @return list<array|string|float|int|bool|null>
```

alexandre-daubois commented:
Thank you Mathias, that was really relevant 👍 A few updates:
- Renamed `JsonPathTokenReducer` to `JsonPathUtils`, inspired by some other classes like `BinaryUtil` from Uid
- Supports both string and resources, with an opt-in dependency to JsonEncoder in case of a resource
- Better path/JSON simplification logic using only iterators and no more arrays as suggested, to leverage the most JsonEncoder's Splitter

alexandre-daubois commented:
Just pushed a fix for the `search()` function that tries to match a regex against any substring of the value.

fabpot commented:
@alexandre-daubois You can rebase this one to use the new `JsonStreamer` component :)

alexandre-daubois commented:
Updated with JsonStreamer, thank you!

chalasr commented on src/Symfony/Component/JsonPath/JsonCrawlerInterface.php:
```diff
@@ -0,0 +1,25 @@
+    /**
+     * @return list<array|string|float|int|bool|null>
+     */
+    public function find(string|JsonPath $query): array;
```
What about errors?

alexandre-daubois replied:
I'm not sure to get which errors you have in mind?

chalasr replied:
I mean the exceptions that may be thrown by implementations starting by the reference one that is shipped in this PR. Should we document them with `@thows ExceptionInterface` or more specific?

alexandre-daubois replied:
You're right, I added `@throws`s with some explanation. Is it better now?

chalasr commented on src/Symfony/Component/JsonPath/Exception/InvalidInputJsonException.php:
```diff
@@ -0,0 +1,27 @@
+namespace Symfony\Component\JsonPath\Exception;
+
+/**
+ * Thrown when a JSON passed as an input is invalid, e.g. in {@see JsonCrawler}.
```
A JSON what? Just "a JSON" doesn't look right to me, like "a PHP".

alexandre-daubois replied:
Indeed, I rephrased this

chalasr commented on src/Symfony/Component/JsonPath/Exception/InvalidInputJsonException.php:
```diff
@@ -0,0 +1,27 @@
+{
+    public function __construct(string $message, ?\Throwable $previous = null)
+    {
+        parent::__construct(\sprintf('Invalid input JSON: %s.', $message), previous: $previous);
```
```suggestion
        parent::__construct(\sprintf('Invalid JSON input: %s.', $message), previous: $previous);
```

chalasr commented on src/Symfony/Component/JsonPath/Exception/InvalidInputJsonException.php:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @experimental
+ */
+class InvalidInputJsonException extends InvalidArgumentException
```
```suggestion
class InvalidJsonInputException extends InvalidArgumentException
```

alexandre-daubois replied:
Renamed to `InvalidJsonStringInputException` per your comment above

alexandre-daubois commented:
Rebased on 7.3.

Status: Needs Review

fabpot commented on src/Symfony/Component/JsonPath/README.md:
```diff
@@ -0,0 +1,37 @@
+==================
+
+The JsonPath component eases JSON navigation using the JSONPath syntax as described in [RFC 9535](https://www.rfc-editor.org/rfc/rfc9535.html).
+
```
Can you add the following:

```
**This Component is experimental**.
[Experimental features](https://symfony.com/doc/current/contributing/code/experimental.html)
are not covered by Symfony's
[Backward Compatibility Promise](https://symfony.com/doc/current/contributing/code/bc.html).
```

alexandre-daubois replied:
Indeed, that's updated now

nicolas-grekas commented on composer.json:
```diff
@@ -84,6 +84,7 @@
         "symfony/http-kernel": "self.version",
         "symfony/intl": "self.version",
         "symfony/json-streamer": "self.version",
+        "symfony/json-path": "self.version",
```
alpha order FTW

alexandre-daubois replied:
Betrayed by JsonStreamer's name change 😄 Here you go

fabpot commented on src/Symfony/Component/JsonPath/JsonCrawlerInterface.php:
```diff
@@ -0,0 +1,31 @@
+{
+    /**
+     * @throws InvalidArgumentException When the JSON string provided to the crawler cannot be decoded
+     * @throws JsonCrawlerException When a syntax error occurs in the provided JSON path
```
Should be moved to after @return

nicolas-grekas reviewed (approved):
Thinking loud about the scope and the name of the component: would it be the place to support navigating hypermedia APIs?

fabpot commented:
Thank you @alexandre-daubois.

stof commented on src/Symfony/Component/JsonPath/JsonCrawler.php:
```diff
@@ -0,0 +1,418 @@
+
+            if (\is_resource($this->raw)) {
+                if (!class_exists(Splitter::class)) {
+                    throw new \LogicException('The JsonEncoder package is required to evaluate a path against a resource. Try running "composer require symfony/json-streamer".');
```
this error message should be updated to talk about the `JsonStreamer` package

alexandre-daubois replied:
Indeed, I'll handle it. Thank you for pointing out!
